rfd/0166-tpm-joining.md

```diff
@@ -0,0 +1,390 @@
+// ProvisionTokenSpecV2
+message ProvisionTokenSpecV2TPM {
+  message Rule {
+    // Hint is a human-readable hint for the rule. This does not impact joining
+    // but can be used for associating a business name with an EKPub or EKCert
+    // serial.
+    // Example: "build-server-100"
+    string Hint = 1  [(gogoproto.jsontag) = "hint,omitempty"];
+    oneof Identifier {
+      // EKPubHash is the SHA256 hash of the EKPub.
+      string EKPubHash = 2 [(gogoproto.jsontag) = "ekpub_hash,omitempty"];
+      // EKCertSerial is the serial number of the EKCert.
+      string EKCertSerial = 3 [(gogoproto.jsontag) = "ekcert_serial,omitempty"];
+    }
+  }
+  // Allow is a list of TokenRules, nodes using this token must match one
+  // allow rule to use this token.
+  repeated Rule Allow = 1 [(gogoproto.jsontag) = "allow,omitempty"];
+  // EKCertAllowCA is a EKCert CA in PEM format.
+  // If present, only TPM devices that present an EKCert that is signed by the
+  // CA specified here may join.
+  //
+  // If not present, then the CA of TPM EKCerts will not be checked during 
+  // joining.
+  bytes EKCertAllowedCA = 2 [(gogoproto.jsontag) = "ekcert_allowed_ca,omitempty"];
+}
+```
+
+In YAML, this would look like:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: my-bot-token
+spec:
+  roles: ["Bot"]
+  bot_name: my-bot
+  join_method: tpm
+  tpm:
+    ekcert_allowed_ca: |
+      -----BEGIN CERTIFICATE-----
+      ... CA obtained from manufacturer ...
+      -----END CERTIFICATE-----
+    allow:
+    - hint: "kansas-build-server-100"
+      ekpub_hash: fc5a1047f5919892fcdf8aa79ea5d6bb6531b5c176939ef0110906cb225941c1
+    - hint: "london-build-server-42"
+      ekcert_serial: 73:df:dc:bd:af:ef:8a:d8:15:2e:96:71:7a:3e:7f:a4
+```
+
+### `RegisterUsingTPMMethod` RPC
+
+A new streaming RPC will be added to the existing `JoinService`:
+
+```protobuf
+service JoinService {
+  // .. Existing Methods ..
+  // RegisterUsingTPMMethod is used to register a Bot or Agent using a TPM.
+  rpc RegisterUsingTPMMethod(stream RegisterUsingTPMMethodRequest) returns (stream RegisterUsingTPMMethodResponse);
+}
+
+// The enrollment challenge response containing the solution returned by
+// calling the TPM2.0 `ActivateCredential` command on the client with the
+// parameters provided in `RegisterUsingTPMMethodEnrollChallengeRequest`.
+message RegisterUsingTPMMethodEnrollChallengeResponse {
+  // The client's solution to `TPMEncryptedCredential` included in
+  // `RegisterUsingTPMMethodEnrollChallengeRequest` using ActivateCredential.
+  bytes solution = 1;
+}
+
+// The attestation key and the parameters necessary to remotely verify it as
+// related to the endorsement key.
+// See https://pkg.go.dev/github.com/google/go-attestation/attest#AttestationParameters.
+// This message excludes the `UseTCSDActivationFormat` field from the link above
+// as it is TMP 1.x specific and always false.
+message TPMAttestationParameters {
```

Comment:
Do we have any controls against replay attacks?  It seems like we should have a timestamp to limit the time window of the Attestation.  The same should be considered for the challenge response.

Reply:
Fortunately we do - this is built into the CredentialActivation ceremony that's built on these parameters. A random challenge (32 bytes) is generated server side and encrypted for the public key of the EK, as part of the credential activation, the TPM then decrypts this secret and this is returned to the server, which validates the secret matches the one it has generated.

Reply:
Fantastic, thank you!